iroha/src/kura.rs

```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
         }
         Ok(BlockStore {
             path: path.to_path_buf(),
+            blocks_per_file,
         })
     }
 
-    fn get_block_filename(block_height: u64) -> String {
-        format!("{}", block_height)
+    /// Returns expected filename of datafile containing block, provided its height
+    async fn get_block_filename(
+        dir: &Path,
+        block_height: NonZeroU64,
+        blocks_per_file: NonZeroU64,
+    ) -> Result<String> {
+        const INITIAL_DATA_FILE: &str = "1";
+        let storage_indices = sorted_storage_files_base_indices(dir)
+            .await
+            .wrap_err("Failed to list storage files")?;
+        match storage_indices.last() {
+            Some(max_base) => {
+                if block_height <= *max_base {
+                    return Err(eyre::eyre!("An attempt to random-access persistent storage (file except for the last one queued)."));
+                }
+                if (block_height.get() - max_base.get()) < blocks_per_file.get() {
+                    return Ok(max_base.to_string());
+                }
+                Ok(block_height.to_string())
+            }
+            None => Ok(String::from(INITIAL_DATA_FILE)),
+        }
     }
 
     /// Get filysystem path for the block at height.
-    pub fn get_block_path(&self, block_height: u64) -> PathBuf {
-        self.path.join(BlockStore::get_block_filename(block_height))
+    ///
+    /// # Errors
+    /// - Filesystem access failure (HW or permissions)
+    ///
+    pub async fn get_block_path(&self, block_height: NonZeroU64) -> Result<PathBuf> {
+        let filename =
+            BlockStore::get_block_filename(&self.path, block_height, self.blocks_per_file)
+                .await
+                .wrap_err("Failed to fetch storage name for block")?;
+        Ok(self.path.join(filename))
     }
 
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .ok_or_else(|| eyre::eyre!("Block height is zero."))?,
+            )
+            .await?;
+        let mut file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
         let serialized_block: Vec<u8> = block.encode_versioned()?;
+        #[allow(clippy::cast_possible_truncation)]
+        let block_size = serialized_block.len() as u32;
+        file.write_all(&block_size.to_ne_bytes())
+            .await
+            .wrap_err("Failed to write to storage file.")?;
```

Comment:
I guess [`AsyncWriteExt::write_u32_le`](https://docs.rs/tokio/1.12.0/tokio/io/trait.AsyncWriteExt.html#method.write_u32_le) suits better here.

Reply:
I don't see a way to apply write_u32_le to File directly - only to intermediate buffer. What's the point then?

Reply:
I mean you write length to file, you can do it by `write_u32_le`.